### PR TITLE
PT-346/PT-349: Make coverage widget configurable

### PR DESCRIPTION
**Pull request by florianmski**
> Tracked in JIRA by [PT-349](https://novoda.atlassian.net/browse/PT-349)

### Scope of the PR

`coverage.js` is now configurable and take an  array of project key as input.
Everything is tested and the code has been a bit cleaned and is now more efficient (get the data for the project currently displayed instead of querying for every projects every time)

Example on how to define the array in `config.json`:

```
{
  [...]
  "services": {
    [...]
  },
  "widgets": {
     [...]
    "coverage": {
      "projects": ["project1", "project2"]
    }
  }
}
```

With `project1` and `project2` being keys you can find on the sonar interface.

### CODE REVIEW

File: node-server/server/widget/coverage.js [hunk 1]
```diff
@@ -1,51 +1,29 @@
...
+    this.httpClient = httpClient || require('request-promise-native');
+    this.projects = widgetConfig.projects || [];
   }
 
   rule() {
-    return httpClient(this._createProjectRequest(this.token))
-      .then(this._parseProjects)
-      .then(this._toAllCoverage(this.token).bind(this))
+    this._incrementIndex();
+    if(this.projects.length == 0) {
```
Comment (frapontillo) [inline node-server/server/widget/coverage.js]: Space after `if`?
I will add this as an ESLint rule.

File: node-server/server/widget/coverage.js [hunk 2]
```diff
@@ -62,28 +40,22 @@ class Coverage extends BaseWidget {
...
+      const projectName = jsonBody.component.name;
+      const measures = component.measures;
       if (measures && measures.length > 0) {
         return Promise.resolve({
           project: projectName,
           coverage: measures[0].value
         });
       } else {
-        return Promise.resolve({});
+        return Promise.reject(`No measures available for ${projectName}`);
```
Comment (frapontillo) [inline node-server/server/widget/coverage.js]: 💯

File: node-server/server/widget/coverage.spec.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  let servicesConfig, widgetConfig;
+  let responsesQueue;
+
+  beforeEach(function() {
+    servicesConfig = {sonar: {token: 'token'}};
+    responsesQueue = [];
+  });
+
+  function httpClient(request) {
```
Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: Isn't `request` unused?

File: node-server/server/widget/coverage.spec.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      .then(result => expect(result).toEqual({
+        widgetKey: 'coverage',
+        template: 'classic',
+        payload: {
+          title: '&#128172;',
+          text: '<b>sun-mobile-nightly</b> has 10.5% coverage'
+        }
+      }))
+      .then(function() {
+        coverage.rule()
```
Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: Why are you nesting? Promises allow you to chain instead:

```javascript
.then(coverage.rule)
.then(result => expect(result).toEqual({
// etc.
```

Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: Have you tried to explicitly call `coverage.rule()` in a function instead of passing the reference?

Comment (florianmski) [inline node-server/server/widget/coverage.spec.js]: Yeah it worked... Any idea why?

File: node-server/server/widget/coverage.spec.js [hunk 5]
```diff
@@ -144,18 +144,16 @@ describe('Coverage', () => {
...
-            widgetKey: 'coverage',
-            template: 'classic',
-            payload: {
-              title: '&#128172;',
-              text: '<b>project2</b> has 99.95% coverage'
-            }
-          }))
-          .then(done);
-      });
+      .then(() => coverage.rule())
```
Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: But you're not returning anything here now... Hmm...

Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: Oh it's implicit!

File: node-server/server/widget/coverage.spec.js (hunk 4 above)
Comment (frapontillo) [inline node-server/server/widget/coverage.spec.js]: I guess [this SO answer](http://stackoverflow.com/a/34930859/801065) will help.

@JozefCeluch and @eduardb for reference.

Comment (JozefCeluch) [inline node-server/server/widget/coverage.spec.js]: @frapontillo do you have any non-SO sources that would explain how context binding works when you pass function references?

**Pull request merged**